bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py

```diff
@@ -12,14 +12,56 @@
 
 import logging
 
+from functools import singledispatch
+from typing import Any, List, Union
+
 import ROOT
 
 from DistRDF.CppWorkflow import CppWorkflow
+from DistRDF.Node import Node
+from DistRDF.Operation import Action, InstantAction, Operation
 from DistRDF.PythonMergeables import SnapshotResult
 
+
 logger = logging.getLogger(__name__)
 
 
+@singledispatch
+def append_node_to_actions(operation: Operation, node: Node, actions: List[Node]) -> None:
+    """
+    Appends the input node to the list of action nodes, if the operation of the
+    node is an action.
+    """
+    pass
+
+
+@append_node_to_actions.register(Action)
+@append_node_to_actions.register(InstantAction)
+def _(operation: Union[Action, InstantAction], node: Node, actions: List[Node]) -> None:
+    actions.append(node)
+
+
+@singledispatch
+def append_node_to_results(operation: Operation, promise: Any, results: list) -> None:
+    """
+    Appends the input promise to the list of results gathered while creating the
+    computation graph, if the operation is an action. The promise can be of many
+    types, usually a 'ROOT.RDF.RResultPtr'. Exceptions are the 'AsNumpy'
+    operation which promise is an 'AsNumpyResult' and the 'Snapshot' operation
+    for which a 'SnapshotResult' is created and appended to the list of results.
+    """
+    pass
+
+
+@append_node_to_results.register(Action)
```

Comment:
So here we need these two decorators plus the Union in the function's signature?

Reply:
Only the two decorators, since `singledispatch` can't use the Union to infer that this overload accepts either an `Action` or an `InstantAction`. The type hint in the argument list is just for style completeness. If you feel it's too much I can remove it

Reply:
Ok, it looks a bit redundant, but might be better for consistency.